File: 0241-different-ways-to-add-parentheses/0241-different-ways-to-add-parentheses.py
class Solution:
    def solve(self, s: str) -> list[int]:
        res = []
        
        for i in range(len(s)):
            c = s[i]
            
            if c in "+-*":
                left = self.solve(s[:i])
                right = self.solve(s[i+1:])
                
                for l in left:
                    for r in right:
                        if c == '+':
                            res.append(l + r)
                        elif c == '-':
                            res.append(l - r)
                        else:  # c == '*'
                            res.append(l * r)
        

        if not res:
            res.append(int(s))
        
        return res

    def diffWaysToCompute(self, expression: str) -> list[int]:
        return self.solve(expression) 

File: 0241-different-ways-to-add-parentheses/0241-different-ways-to-add-parentheses.cpp
class Solution {
public:
    vector<int>solve(string s){
        vector<int>res;
        for(int i=0; i<s.length();i++){
           char c=s[i];
           if(c=='+' || c=='-' || c=='*'){
              vector<int>left = solve(s.substr(0,i));
              vector<int>right= solve(s.substr(i+1));
              for(auto &l:left){
                for(auto &r:right){
                    (c=='+')?res.push_back(l+r):((c=='-')?res.push_back(l-r):res.push_back(l*r));
                    }
                  }
                }
            }
        if(res.empty())res.push_back(stoi(s));
        return res;
    }
    vector<int> diffWaysToCompute(string expression) {
        return solve(expression);
    }
};